src/Storages/MergeTree/MergeTreeIndexSet.cpp

```diff
@@ -318,17 +333,214 @@ bool MergeTreeIndexConditionSet::mayBeTrueOnGranule(MergeTreeIndexGranulePtr idx
+    ActionsDAGPtr & result_dag,
+    const ContextPtr & context,
+    std::unordered_map<const ActionsDAG::Node *, const ActionsDAG::Node *> & node_to_result_node) const
+{
+    /// Functions AND, OR, NOT. Replace with bit*.
+
+    const auto * node_to_check = &node;
+    while (node_to_check->type == ActionsDAG::ActionType::ALIAS)
+        node_to_check = node_to_check->children[0];
+
+    if (node_to_check->column && isColumnConst(*node_to_check->column))
+        return nullptr;
+
+    if (node_to_check->type != ActionsDAG::ActionType::FUNCTION)
+        return nullptr;
+
+    auto function_name = node_to_check->function->getName();
+    const auto & arguments = node_to_check->children;
+    size_t arguments_size = arguments.size();
+
+    if (function_name == "not")
+    {
+        if (arguments_size != 1)
+            return nullptr;
+
+        auto bit_swap_last_two_function = FunctionFactory::instance().get("__bitSwapLastTwo", context);
+        return &result_dag->addFunction(bit_swap_last_two_function, {arguments[0]}, {});
+    }
+    else if (function_name == "and" || function_name == "indexHint" || function_name == "or")
+    {
+        if (arguments_size < 2)
+            return nullptr;
+
+        ActionsDAG::NodeRawConstPtrs children;
+        children.resize(arguments_size);
+
+        for (size_t i = 0; i < arguments_size; ++i)
+            children[i] = &traverseDAG(*arguments[i], result_dag, context, node_to_result_node);
+
+        FunctionOverloadResolverPtr function;
+
+        if (function_name == "and" || function_name == "indexHint")
+            function = FunctionFactory::instance().get("__bitBoolMaskAnd", context);
+        else
+            function = FunctionFactory::instance().get("__bitBoolMaskOr", context);
+
+        const auto * last_argument = children.back();
+        children.pop_back();
+
+        const auto * before_last_argument = children.back();
+        children.pop_back();
+
+        while (true)
+        {
+            last_argument = &result_dag->addFunction(function, {before_last_argument, last_argument}, {});
+
+            if (children.empty())
+                break;
+
+            before_last_argument = children.back();
+            children.pop_back();
+        }
+
+        return last_argument;
+    }
+
+    return nullptr;
+}
+
+bool MergeTreeIndexConditionSet::checkDAGUseless(const ActionsDAG::Node & node, const ContextPtr & context, bool atomic) const
+{
+    const auto * node_to_check = &node;
+    while (node_to_check->type == ActionsDAG::ActionType::ALIAS)
+        node_to_check = node_to_check->children[0];
+
+    RPNBuilderTreeContext tree_context(context);
+    RPNBuilderTreeNode tree_node(node_to_check, tree_context);
+
+    if (node.column && isColumnConst(*node.column))
+    {
+        Field literal;
+        node.column->get(0, literal);
+        return !atomic && literal.safeGet<bool>();
```

Comment:
A bit tricky line. We check if the literal is a bool and do not read it's value. Maybe add a comment?